### PR TITLE
User notifications

### PR DESCRIPTION
**Pull request by acrellin**
This PR implements front-end user notifications. A notifications badge is added to the top bar that, when clicked, displays all of the user's notifications in a scrollable popover. They can then be marked read, deleted, or, if applicable, clicked to redirect to the relevant page. 

Users are notified when:
 - they're mentioned in a comment;
 - a group for which they're the first admin member receives a new admission request;
 - they are added to a new group;
 - their group admission request is accepted/declined

Comments are parsed for username tags (containing "@username").

Sample screen captures:
comment_mention_notification

group_admission_request_acceptance_notification

added_to_group_notification

group_admission_request_notification

Closes https://github.com/skyportal/skyportal/issues/1481

### CODE REVIEW

**Comment by acrellin** [comment]
Test failures (2) are now unrelated: 

```
FAILED skyportal/tests/frontend/test_followup_requests.py::test_delete_followup_request_SEDM
FAILED skyportal/tests/frontend/test_sources.py::test_regular_user_cannot_delete_unowned_comment
= 2 failed, 349 passed, 3 xfailed, 1 xpassed, 1 warning, 12 rerun in 2207.54s (0:36:47) =
```

File: skyportal/handlers/api/group_admission_request.py [hunk 1]
```diff
@@ -138,19 +140,38 @@ def post(self):
...
+        group_admin = (
+            User.query.get(group_admin_gu.user_id)
+            if group_admin_gu is not None
+            else None
+        )
+        if group_admin is not None:
+            DBSession().add(
+                UserNotification(
+                    user=group_admin,
+                    text=f"{requesting_user.username} has requested to join {group.name}",
```
Comment (dmitryduev) [inline skyportal/handlers/api/group_admission_request.py:167]: [Chips](https://material-ui.com/components/chips/#chip-2)?

File: skyportal/handlers/api/group_admission_request.py [hunk 2]
```diff
@@ -201,7 +222,15 @@ def patch(self, admission_request_id):
         ):
             return self.error("Insufficient permissions.")
         admission_request.status = status
+        DBSession().add(
+            UserNotification(
+                user=admission_request.user,
+                text=f"Your admission request to group {admission_request.group.name} has been {status}.",
```
Comment (dmitryduev) [inline skyportal/handlers/api/group_admission_request.py:228]: [Chips](https://material-ui.com/components/chips/#chip-2)?

File: skyportal/tests/frontend/test_notifications.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    comment_box = driver.wait_for_xpath("//input[@name='text']")
+    comment_text = f"@{user.username}"
+    comment_box.send_keys(comment_text)
+    driver.click_xpath('//*[@name="submitCommentButton"]')
+    driver.wait_for_xpath(f'//p[text()="{comment_text}"]')
+
+    driver.wait_for_xpath("//span[text()='1']")
+    driver.click_xpath('//*[@data-testid="notificationsButton"]')
+    driver.wait_for_xpath(
+        f'//*[text()="{user.username} mentioned you in a comment on {public_source.id}"]'
```
Comment (dmitryduev) [inline skyportal/tests/frontend/test_notifications.py:21]: Chips

File: skyportal/tests/frontend/test_notifications.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    )
+    driver.click_xpath('//*[contains(@data-testid, "markReadButton")]')
+    driver.wait_for_xpath_to_disappear("//span[text()='1']")
+    driver.click_xpath('//*[contains(@data-testid, "deleteNotificationButton")]')
+    driver.wait_for_xpath_to_disappear(
+        f'//*[text()="{user.username} mentioned you in a comment on {public_source.id}"]'
```
Comment (dmitryduev) [inline skyportal/tests/frontend/test_notifications.py:27]: Chips

**Review by dmitryduev — CHANGES_REQUESTED** [review]
Looks really awesome @acrellin! I have a few small requests, all on the UI/UX side.

File: static/js/components/Notifications.jsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+  return (
+    <>
+      <IconButton
+        onClick={handleClickOpen}
+        style={{ marginTop: "0.6rem", marginRight: "1rem" }}
+        data-testid="notificationsButton"
+      >
+        <Badge
+          badgeContent={unreadCount}
```
[the remark this answers is not in the document]
Comment (acrellin) [reply static/js/components/Notifications.jsx:91]: 👍🏼

**Comment by acrellin** [comment]
@dmitryduev thanks for the review. Regarding styling the message contents w/ Chips, the text is all generated on the Python side. We could possible search for things like "mentioned you", and then split out the author's username, etc, but it'd be a bit of a pain and would easily break if we even slightly change the message format on the back-end.

**Comment by acrellin** [comment]
@dmitryduev unless you know of a simple, non-hacky way to do that?

File: static/js/components/Notifications.jsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+  const deleteNotification = (notificationID) => {
+    dispatch(userNotificationsActions.deleteNotification(notificationID));
+  };
+
+  return (
+    <>
+      <IconButton
+        onClick={handleClickOpen}
+        style={{ marginTop: "0.6rem", marginRight: "1rem" }}
```
[the remark this answers is not in the document]
Comment (acrellin) [reply static/js/components/Notifications.jsx]: Done

**Comment by dmitryduev** [comment]
@acrellin ah, right, makes sense. Can't think of a clean/easy way... @kmshin1397 can you?

**Comment by stefanv** [comment]
How about parsing the message and inserting a special Markdown link:

`[@mention](user:id)`

Or something like that.

**Comment by dmitryduev** [comment]
right @stefanv! we can parse md on the frontend, can we? Backticks would also do then.

**Comment by stefanv** [comment]
I thought we rendered markdown on the frontend.  If we don't already, it's probably worth adding :)  There are several react components for that.

**Comment by kmshin1397** [comment]
Yeah I like Stefan's suggestion of using some special syntax to denote highlighted segments. Maybe we can tag these with URLs so that click on the chips would take you to the group page etc? Like the `[text](url)` markdown link syntax. And just render what's in the brackets as Chip text with a link to `url`. Although we don't really have a useful User page so I guess the username chips wouldn't be all the useful.

**Comment by acrellin** [comment]
@dmitryduev 

mark_notifs_read_unread

**Comment by acrellin** [comment]
I would like for parsing markdown / styling tags to be added in a later PR. This is the core notifications functionality.

**Comment by acrellin** [comment]
And I nominate one of you who has a clear idea of how it should be done to tackle that issue ;)

File: static/js/components/Notifications.jsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                  </ListItem>
+                  <ListItem className={classes.centered}>
+                    <Button
+                      data-testid={`markReadButton${notification.id}`}
+                      size="small"
+                      onClick={() => {
+                        markRead(notification.id);
+                      }}
+                    >
+                      Mark read
```
[the remark this answers is not in the document]
Comment (acrellin) [reply static/js/components/Notifications.jsx]: Done

**Comment by stefanv** [comment]
I don't think this PR needs that functionality, Ari.  Was just throwing some ideas out there.

**Comment by acrellin** [comment]
> I don't think this PR needs that functionality, Ari. Was just throwing some ideas out there.

Yeah I think that was clear, thanks, @stefanv :)  I was just making it clear to Dima that he should re-review now and we can add the highlighting stuff later

**Review by dmitryduev — APPROVED** [review]
👍🏻

**Comment by dmitryduev** [comment]
Markdown parsing definitely should not be part of this PR:)

**Comment by acrellin** [comment]
Thanks for the review and great feedback, @dmitryduev

**Review by kmshin1397 — CHANGES_REQUESTED** [review]
Looks good, @acrellin! I would suggest adding tests for the group membership notifications as well. Once that's in place I think this is good to merge.

**Comment by acrellin** [comment]
@kmshin1397 I've added a test that ensures that both the group admin receives a notification when someone requests admission to a group, and that the requesting user receives a notification when their request has been processed. Ready for re-review

**Review by kmshin1397 — COMMENTED** [review]
Awesome - sorry, two more minor suggestions

**Comment by acrellin** [comment]
@kmshin1397 👍🏼 Done

**Review by kmshin1397 — APPROVED** [review]

**Comment by dmitryduev** [comment]
Can't wait to roll this out! Behold, Slack, Fritz is coming :))

**Comment by acrellin** [comment]
Failures unrelated

**Pull request merged**